airflow/jobs/scheduler_job.py

```diff
@@ -534,42 +535,63 @@ def update_import_errors(session: Session, dagbag: DagBag) -> None:
+        :param dagbag: Dag Bag of dags
+        :param callback_requests: failure callbacks to execute
+        :type callback_requests: List[airflow.utils.callback_requests.CallbackRequest]
         :param session: DB session.
         """
-        for request in failure_callback_requests:
-            simple_ti = request.simple_task_instance
-            if simple_ti.dag_id in dagbag.dags:
-                dag = dagbag.dags[simple_ti.dag_id]
-                if simple_ti.task_id in dag.task_ids:
-                    task = dag.get_task(simple_ti.task_id)
-                    ti = TI(task, simple_ti.execution_date)
-                    # Get properties needed for failure handling from SimpleTaskInstance.
-                    ti.start_date = simple_ti.start_date
-                    ti.end_date = simple_ti.end_date
-                    ti.try_number = simple_ti.try_number
-                    ti.state = simple_ti.state
-                    ti.test_mode = self.UNIT_TEST_MODE
-                    ti.handle_failure(request.msg, ti.test_mode, ti.get_template_context())
-                    self.log.info('Executed failure callback for %s in state %s', ti, ti.state)
+        for request in callback_requests:
+            if isinstance(request, TaskCallbackRequest) and request.is_failure_callback:
+                self._execute_task_callbacks(dagbag, request)
+            elif isinstance(request, SlaCallbackRequest):
+                self.manage_slas(dagbag.dags.get(request.dag_id))
+            elif isinstance(request, DagCallbackRequest):
+                self._execute_dag_callbacks(dagbag, request, session)
+
         session.commit()
 
+    @provide_session
+    def _execute_dag_callbacks(self, dagbag: DagBag, request: DagCallbackRequest, session: Session):
+        dag = dagbag.dags[request.dag_id]
+        dag_run = dag.get_dagrun(execution_date=request.execution_date, session=session)
+        dag.handle_callback(
+            dagrun=dag_run,
+            success=not request.is_failure_callback,
+            reason=request.msg,
+            session=session
+        )
+
+    def _execute_task_callbacks(self, dagbag: DagBag, request: TaskCallbackRequest):
+        simple_ti = request.simple_task_instance
+        if simple_ti.dag_id in dagbag.dags:
+            dag = dagbag.dags[simple_ti.dag_id]
+            if simple_ti.task_id in dag.task_ids:
+                task = dag.get_task(simple_ti.task_id)
+                ti = TI(task, simple_ti.execution_date)
+                # Get properties needed for failure handling from SimpleTaskInstance.
+                ti.start_date = simple_ti.start_date
+                ti.end_date = simple_ti.end_date
+                ti.try_number = simple_ti.try_number
+                ti.state = simple_ti.state
+                ti.test_mode = self.UNIT_TEST_MODE
+                ti.handle_failure(request.msg, ti.test_mode, ti.get_template_context())
```

Comment:
Isn't this happening in the main thread still?

Reply:
Oh, wrong class, this is inside DagFileProcessor, right?

Reply:
Yup